kumir/KumirLexer.g4

```diff
@@ -0,0 +1,154 @@
+
+// --- Operators ---
+POWER               : '**'; // Power operator
+GE                  : '>=' | '≥'; // Greater than or equal
+LE                  : '<=' | '≤'; // Less than or equal
+NE                  : '<>' | '≠'; // Not equal
+PLUS                : '+';
+MINUS               : '-';
+MUL                 : '*';
+DIV                 : '/';
+EQ                  : '='; // Equal (used in comparisons and initialization)
+LT                  : '<';
+GT                  : '>';
+LPAREN              : '('; // Left parenthesis
+RPAREN              : ')'; // Right parenthesis
+LBRACK              : '['; // Left square bracket (arrays)
+RBRACK              : ']'; // Right square bracket (arrays)
+LBRACE              : '{'; // Left curly brace (array literals)
+RBRACE              : '}'; // Right curly brace (array literals)
+COMMA               : ','; // Comma
+COLON               : ':'; // Colon (array bounds, output format)
+SEMICOLON           : ';'; // Semicolon (statement separator)
+ATAT                : '@@'; // Double at-sign (special teacher functions)
+AT                  : '@';  // At-sign (used in identifiers)
+
+// --- Literals ---
+// Character literal: single character in single quotes
+CHAR_LITERAL        : '\'' ( EscapeSequence | ~['\\\r\n] ) '\'' ;
+// String literal: sequence of characters in double or single quotes
+STRING              : '"' ( EscapeSequence | ~["\\\r\n] )*? '"'
+                    | '\'' ( EscapeSequence | ~['\\\r\n] )*? '\''
+                    ;
+// Real literal: with decimal point or in exponential form
+REAL                : (DIGIT+ '.' DIGIT* | '.' DIGIT+) ExpFragment? // 123. , .5 , 123.45
+                    | DIGIT+ ExpFragment                          // 123e4
+                    ;
+// Integer literal: decimal or hexadecimal (starts with $)
+INTEGER             : DecInteger | HexInteger ;
+
+// --- Identifier ---
+// Variable, algorithm, module name, etc.
+// Starts with a letter, followed by letters, digits, '_', or '@'.
+ID                  : LETTER (LETTER | DIGIT | '_' | '@')* ;
+
+// --- Comments ---
+// Comments starting with '|' or '#' are ignored by the parser.
+// The comment text is sent to the HIDDEN channel.
+// Does NOT consume the trailing newline.
+LINE_COMMENT        : '|' ~[\r\n]* -> channel(HIDDEN); // Single-line comment
+DOC_COMMENT         : '#' ~[\r\n]* -> channel(HIDDEN); // Documentation comment
+
+// --- Whitespace ---
+// Spaces, tabs, newlines are skipped by the lexer.
+WS                  : [ \t\r\n]+ -> skip;
+
+// --- Fragments ---
+// Fragments are helper rules for the lexer and do not produce tokens themselves.
+fragment DIGIT      : [0-9];
+fragment HEX_DIGIT  : [0-9a-fA-F];
+// Letter (Latin or Cyrillic, including ё/Ё)
+fragment LETTER     : [a-zA-Zа-яА-ЯёЁ];
+fragment DecInteger : DIGIT+;
+fragment HexInteger : '$' HEX_DIGIT+;
+// Exponential part of a real number (e.g., e+10, E-5, е-3)
+fragment ExpFragment: [eEеЕ] [+-]? DIGIT+;
```

Comment:
Duplicated literals (actually only single one is required if use `caseInsensitive` option):

```suggestion
fragment ExpFragment: [eE] [+-]? DIGIT+;
```